# **I. Airline Dataset Audit**

## **1. Introduction & Objectives**

This section records the initial screening of the **2015 Flight Delays and Cancellations** dataset and the supplementary data-quality checks completed after the P0 write benchmark.
The review documents dataset dimensions, inferred schema, missing values, and the fields used by the downstream read-query and partitioning experiments.

**What will be audited:**
- Physical file size on disk.
- Total row and column counts.
- Schema and datatypes across the dataset.
- Missing values (Nulls) distribution.
- Validation of four critical columns required for Task 2:
  - `YEAR` and `MONTH` (Categorical columns for `.partitionBy()` testing).
  - `AIRLINE` and `ARRIVAL_DELAY` (grouping by `AIRLINE` and computing `AVG(ARRIVAL_DELAY)`).
- Distinct values and distributions relevant to the planned grouping and partitioning experiments.

**Scope limitations:**
This notebook is strictly for **Data Auditing**. It does **not** perform:
- Data cleaning or dropping rows.
- Filling missing values.
- Feature engineering or renaming columns.
- Format conversions or storage benchmarking (which will be handled in the main benchmark scripts).

In [ ]:
# ============================================================
# AIRLINE DATASET AUDIT
# ============================================================

import os
from pyspark.sql import SparkSession, functions as F
from pyspark.storagelevel import StorageLevel

# ------------------------------------------------------------
# 1. Initialize Spark and read the raw dataset
# ------------------------------------------------------------
INPUT_FILE = r"C:\BigDataProject\data\raw\airline\flights.csv"

spark = (SparkSession.builder
    .appName("DatasetCandidateScreening")
    .master("local[*]")
    .config("spark.sql.shuffle.partitions", "8")
    .getOrCreate()
)

spark.sparkContext.setLogLevel("WARN")

airline_df = (spark.read
    .option("header", True)
    .option("inferSchema", True)
    .csv(INPUT_FILE)
    .persist(StorageLevel.MEMORY_AND_DISK)
)

# count() gets the row count and materializes the DataFrame so that
# subsequent audits do not have to re-read the entire CSV from scratch.
row_count = airline_df.count()
column_count = len(airline_df.columns)

# ------------------------------------------------------------
# 2. General information
# ------------------------------------------------------------
print("=" * 70)
print("BASIC DATASET INFORMATION")
print("=" * 70)

print(f"File size    : {os.path.getsize(INPUT_FILE) / (1024**2):,.2f} MiB")
print(f"Row count    : {row_count:,}")
print(f"Column count : {column_count}")
print(f"Spark version: {spark.version}")

print("\nColumns:")
print(airline_df.columns)

print("\nSchema:")
airline_df.printSchema()

print("\nSample rows:")
airline_df.show(5, truncate=False)

# ------------------------------------------------------------
# 3. Missing values of the entire dataset
# ------------------------------------------------------------
null_counts = (
    airline_df.select([
        F.sum(F.col(c).isNull().cast("int")).alias(c)
        for c in airline_df.columns
    ])
    .first()
    .asDict()
)

null_summary = spark.createDataFrame([(
        column,
        int(null_counts[column]),
        round(null_counts[column] / row_count * 100, 4)
    )
    for column in airline_df.columns
], ["column", "null_count", "null_percentage"])

print("\n" + "=" * 70)
print("NULL VALUE SUMMARY")
print("=" * 70)

null_summary.orderBy(F.desc("null_percentage")).show(column_count, truncate=False)

# ------------------------------------------------------------
# 4. Check critical columns for Task 2
# ------------------------------------------------------------
critical_cols = ["YEAR", "MONTH", "AIRLINE", "ARRIVAL_DELAY"]

print("\n" + "=" * 70)
print("CRITICAL COLUMN CHECK")
print("=" * 70)

for c in critical_cols:
    if c in airline_df.columns:
        print(f"[FOUND] {c:<15} -> {dict(airline_df.dtypes)[c]}")
    else:
        print(f"[MISSING] {c}")

# ------------------------------------------------------------
# 5. Statistics of critical columns
# ------------------------------------------------------------
critical_stats = airline_df.agg(
    F.min("YEAR").alias("min_year"),
    F.max("YEAR").alias("max_year"),
    F.countDistinct("YEAR").alias("distinct_years"),

    F.min("MONTH").alias("min_month"),
    F.max("MONTH").alias("max_month"),
    F.countDistinct("MONTH").alias("distinct_months"),

    F.countDistinct("AIRLINE").alias("distinct_airlines"),

    F.min("ARRIVAL_DELAY").alias("min_arrival_delay"),
    F.max("ARRIVAL_DELAY").alias("max_arrival_delay"),
    F.avg("ARRIVAL_DELAY").alias("avg_arrival_delay"),
    F.sum(F.col("ARRIVAL_DELAY").isNull().cast("int"))
     .alias("null_arrival_delay")
)

print("\n" + "=" * 70)
print("CRITICAL COLUMN STATISTICS")
print("=" * 70)

critical_stats.show(truncate=False)

# ------------------------------------------------------------
# 6. Distribution by Month
# ------------------------------------------------------------
print("\n" + "=" * 70)
print("MONTH DISTRIBUTION")
print("=" * 70)

(airline_df.groupBy("MONTH")
    .count()
    .orderBy("MONTH")
    .show(20, truncate=False)
)

# ------------------------------------------------------------
# 7. Distribution by Airline
# ------------------------------------------------------------
print("\n" + "=" * 70)
print("AIRLINE DISTRIBUTION")
print("=" * 70)

(airline_df.groupBy("AIRLINE")
    .count()
    .orderBy(F.desc("count"))
    .show(50, truncate=False)
)

# ------------------------------------------------------------
# 8. End of audit
# ------------------------------------------------------------
airline_df.unpersist()

ERROR:root:Exception while sending command.
Traceback (most recent call last):
  File "c:\BigDataProject\.venv\Lib\site-packages\py4j\clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
                          ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\Lib\socket.py", line 706, in readinto
    return self._sock.recv_into(b)
           ^^^^^^^^^^^^^^^^^^^^^^^
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\BigDataProject\.venv\Lib\site-packages\py4j\java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\BigDataProject\.venv\Lib\site-packages\py4j\clientserver.py", line 539, in send_command
    

Py4JError: An error occurred while calling o33.count

ERROR:py4j.clientserver:Unknown command A


## **Analysis of Results**

### **1. Dataset Scale**

Based on the general information output, we can record the following dimensions:

* **Source file size:** 564.96 MiB.
* **Row count:** 5,819,079.
* **Column count:** 31.

**Interpretation:** These dimensions describe the workload selected for the format benchmark. Differences in storage size and write execution time are evaluated using the recorded benchmark results.
**Unit note:** The historical screening output labels the source size as MB. The calculation divides bytes by `1024**2`, so the displayed value should be interpreted as **564.96 MiB**.

### **2. Schema Validation**

The screening recorded Spark's inferred schema for `flights.csv` and checked the presence and data types of the four fields used by the read-query and partitioning experiments:

| Column | Role in Task 2 | Audit Result |
|---|---|---|
| `YEAR` | Partition column | Validated (`integer`) |
| `MONTH` | Partition column | Validated (`integer`) |
| `AIRLINE` | Categorical column / `GROUP BY` | Validated (`string`) |
| `ARRIVAL_DELAY` | Numeric measure / `AVG()` | Validated (`integer`) |

Since all four critical columns exist with the correct datatypes, this dataset can be mapped almost directly to the canonical schema required by the project:

- `AIRLINE` → `Carrier`
- `ARRIVAL_DELAY` → `ArrDelay`
- `YEAR` → `Year`
- `MONTH` → `Month`

**Schema note:** The original screening reads the source column names. The four renames listed above were subsequently applied in the benchmark pipeline to produce the retained canonical field names.

### **3. Missing Values Summary**

Based on the `NULL VALUE SUMMARY` table, we identified the following data quality conditions:

* **Missingness in cancellation and delay-reason fields:** Several inspected columns, including `CANCELLATION_REASON`, `SECURITY_DELAY`, and `WEATHER_DELAY`, have recorded null rates ranging from approximately **81% to over 98%**. These rates describe missingness; they do not, by themselves, establish that the records are erroneous.
* **Partitioning and grouping columns:** `YEAR`, `MONTH`, and `AIRLINE` have **zero missing values** in the recorded screening output. This establishes completeness for these fields. Value-domain checks and the supplementary carrier-lookup review provide separate evidence about the validity and reference consistency of their recorded values.
* **Metric Column:** `ARRIVAL_DELAY` contains **105,071 missing values** (approximately **1.81%**). The supplementary post-P0 review of the canonical Parquet output found that all missing `ArrDelay` values occur in records marked as cancelled (**89,884**) or diverted (**15,187**). No missing arrival delays were found among the **5,714,008 records** with both status flags equal to zero.

**Retained-data policy:**

* The benchmark preserves all **5,819,079 rows** and the **105,071 missing arrival-delay values**.
* The supplementary review documents the association between arrival-delay availability and the recorded cancellation and diversion flags.
* These findings provide additional interpretation of the retained dataset.

### **4. Year and Month Distribution**

Checking the bounds and cardinality of the temporal columns:

- `min_year`: 2015
- `max_year`: 2015
- `distinct_years`: 1
- `min_month`: 1
- `max_month`: 12
- `distinct_months`: 12

**Conclusion:**

* The recorded values cover **2015** and all **12 months**, giving **12 distinct Year–Month combinations**.
* `YEAR` and `MONTH` are mapped to the canonical `Year` and `Month` fields used by the downstream partitioning experiment.
* **The number of partition-value combinations does not determine the number or size of data files.** The partitioning experiment must inspect the files actually written before drawing conclusions about small files.


### **5. Airline Distribution**

Checking the categorical grouping column:

* The dataset contains **14 distinct carrier codes**.
* **WN** has the largest recorded group, with **1,261,855 rows**.
* **VX** has the smallest recorded group, with **61,903 rows**.

**Interpretation:** `AIRLINE` provides a categorical grouping field for the planned `GROUP BY Carrier` query. The observed counts show unequal group sizes. Their effect on task balance and query execution time requires execution-level evidence.


### **6. Arrival Delay Profile**

Checking the numeric metric column:

* **Minimum arrival delay:** -87 (Early arrivals)
* **Maximum arrival delay:** 1971 (Extreme delays)
* **Mean arrival delay over non-null values:** approximately **4.41**.
* **Null values:** 105,071

**Interpretation:**
* Spark inferred `ARRIVAL_DELAY` as an **integer** column. The observed values range from **−87** to **1,971**, with a non-null mean of approximately **4.41**.
* These summaries describe the recorded numeric values. They do not establish the validity of every individual record.
* The supplementary missingness review documents the **105,071 null values** and their correspondence with the recorded cancellation and diversion flags.
* **All recorded values and nulls are retained under the existing data contract.**

### **7. Conclusion for Candidate A**

Following this audit, we can summarize Candidate A (Airline 2015) against our project criteria:

| Evaluation Criteria | Airline 2015 Dataset |
|---|---|
| Spark successfully reads the dataset | Yes |
| Suitable categorical column exists | Yes (`AIRLINE`, 14 distinct) |
| Suitable numeric column exists | Yes (`ARRIVAL_DELAY`) |
| Native `YEAR` column available | Yes; inferred as integer and mapped to canonical `Year` |
| Native `MONTH` column available | Yes; inferred as integer and mapped to canonical `Month` |
| Arrival-delay missingness | 105,071 null values, approximately 1.81%; retained and described in the supplementary review |
| Needs Year/Month derivation logic | No (Ready to use) |
| Fits project's benchmark query | Yes |
| Notable missing values issues | Minor (1.8% in `ARRIVAL_DELAY`, acceptable) |

**Historical context:** This section records the initial Airline screening. The candidate comparison is retained to explain the original selection.
**Airline 2015 is the retained dataset for the completed write benchmark. **The supplementary reviews document additional findings after P0.

## **Supplementary Data Quality Checks**

This section extends the initial screening with carrier reference checks and an analysis of missing arrival delays.

* Carrier codes will be checked against the airline lookup supplied with the same dataset.
* Missing arrival delays will be examined alongside cancellation and diversion indicators.
* Findings document the quality and interpretation of the retained dataset.

**Execution dependency:** The arrival-delay review reads the existing P0 Parquet output from `output/format_benchmark/parquet/`. Run this review after the accepted benchmark outputs are available.

### **Carrier Lookup Validation**

* Check the completeness and uniqueness of carrier codes in `airlines.csv`.
* Identify surrounding whitespace and unexpected code shapes.
* Compare lookup coverage with the **14 carrier codes recorded in the accepted P0 snapshot**.
* Use the results to document data quality while preserving the existing dataset and benchmark outputs.

This check assesses consistency with the dataset's reference table. It does not independently certify IATA assignments or re-scan the flight data.

In [ ]:
# Reuse the exact carrier codes verified in the accepted P0 snapshot.
p0_carrier_codes = {
    "AA", "AS", "B6", "DL", "EV", "F9", "HA",
    "MQ", "NK", "OO", "UA", "US", "VX", "WN",
}

lookup_codes = carrier_lookup["IATA_CODE"]
lookup_names = carrier_lookup["AIRLINE"]

# Build diagnostic masks without modifying the lookup values.
valid_code_shape = lookup_codes.str.fullmatch(r"[A-Z0-9]{2}", na=False)
code_whitespace = lookup_codes.ne(lookup_codes.str.strip()).fillna(False)
name_whitespace = lookup_names.ne(lookup_names.str.strip()).fillna(False)

# Compare exact values so that formatting differences remain observable.
lookup_code_set = set(lookup_codes.dropna())
missing_p0_codes = sorted(p0_carrier_codes - lookup_code_set)
extra_lookup_codes = sorted(lookup_code_set - p0_carrier_codes)

lookup_issue_counts = {
    "missing_or_blank_codes": lookup_codes.fillna("").str.strip().eq("").sum(),
    "codes_with_outer_whitespace": code_whitespace.sum(),
    "unexpected_code_shape": (~valid_code_shape).sum(),
    "rows_with_duplicate_codes": lookup_codes.duplicated(keep=False).sum(),
    "missing_or_blank_names": lookup_names.fillna("").str.strip().eq("").sum(),
    "names_with_outer_whitespace": name_whitespace.sum(),
    "p0_codes_missing_from_lookup": len(missing_p0_codes),
    "lookup_codes_absent_from_p0": len(extra_lookup_codes),
}

carrier_checks = pd.DataFrame(
    lookup_issue_counts.items(), columns=["check", "issue_count"]
)
carrier_checks["status"] = carrier_checks["issue_count"].eq(0).map(
    {True: "PASS", False: "REVIEW"}
)

print("=== CARRIER LOOKUP VALIDATION ===")
display(carrier_checks)

print("P0 codes missing from lookup:", missing_p0_codes)
print("Lookup codes absent from P0:", extra_lookup_codes)

=== CARRIER LOOKUP VALIDATION ===


,check,issue_count,status
0,missing_or_blank_codes,0,PASS
1,codes_with_outer_whitespace,0,PASS
2,unexpected_code_shape,0,PASS
3,rows_with_duplicate_codes,0,PASS
4,missing_or_blank_names,0,PASS
5,names_with_outer_whitespace,0,PASS
6,p0_codes_missing_from_lookup,0,PASS
7,lookup_codes_absent_from_p0,0,PASS


P0 codes missing from lookup: []
Lookup codes absent from P0: []


#### **Carrier Lookup: Results**

* **All eight checks passed**, with zero issues detected under the specified checks.
* The lookup contains **14 unique carrier codes**, each associated with a nonblank airline name. No surrounding whitespace was detected in either field.
* All codes match the two-character uppercase alphanumeric shape used in this check, including codes such as **B6** and **F9**.
* Lookup coverage matches the **14 carrier codes recorded in the accepted P0 snapshot**. No P0 codes are missing, and no additional lookup codes are present.
* These findings support consistency with the reference table supplied with the dataset. They do not independently verify historical IATA assignments.
* **The existing dataset and benchmark outputs are retained unchanged.**

### **Arrival Delay Missingness: Post-P0 Review**

* Examine missing `ArrDelay` values across combinations of **CANCELLED** and **DIVERTED**.
* Report both the **missing-value rate within each group** and each group's **share of all missing arrival delays**.
* Identify missing flag values and values outside the expected **0/1** domain.
* Read the accepted **Parquet output** as the source for this supplementary analysis. Retain the existing dataset and benchmark outputs unchanged.

This review describes the relationship between recorded flight-status flags and arrival-delay availability.

In [ ]:
from IPython.display import display
from pyspark.sql import SparkSession, functions as F

parquet_path = r"C:\BigDataProject\output\format_benchmark\parquet"
qa_spark = (
    SparkSession.builder.master("local[*]")
    .appName("AirlineDataQuality").getOrCreate()
)
qa_spark.sparkContext.setLogLevel("WARN")

# Label flag states only in the temporary diagnostic view.
flag_labels = [
    F.when(F.col(flag).isNull(), "NULL")
    .when(F.col(flag) == 0, "0")
    .when(F.col(flag) == 1, "1")
    .otherwise("INVALID")
    .alias(flag)
    for flag in ("CANCELLED", "DIVERTED")
]

# Collect only the aggregated table, with at most 16 flag combinations.
delay_groups = (
    qa_spark.read.parquet(parquet_path)
    .select("ArrDelay", *flag_labels)
    .groupBy("CANCELLED", "DIVERTED")
    .agg(
        F.count("*").alias("total_rows"),
        F.count(F.when(F.col("ArrDelay").isNull(), 1)).alias("null_arrdelay"),
        F.count("ArrDelay").alias("nonnull_arrdelay"),
    )
    .orderBy("CANCELLED", "DIVERTED")
    .toPandas()
)

total_rows = int(delay_groups["total_rows"].sum())
total_missing = int(delay_groups["null_arrdelay"].sum())

delay_groups["null_rate_pct"] = (
    100 * delay_groups["null_arrdelay"] / delay_groups["total_rows"]
).round(2)
delay_groups["share_of_nulls_pct"] = (
    (100 * delay_groups["null_arrdelay"] / total_missing).round(2)
    if total_missing else float("nan")
)

print("=== ARRIVAL DELAY MISSINGNESS ===")
print(f"Source: {parquet_path}")
print(f"Rows: {total_rows:,} | P0 match: {total_rows == 5_819_079}")
print(f"Null ArrDelay: {total_missing:,} | P0 match: {total_missing == 105_071}")
display(delay_groups)

for flag in ("CANCELLED", "DIVERTED"):
    null_rows = delay_groups.loc[delay_groups[flag].eq("NULL"), "total_rows"].sum()
    invalid_rows = delay_groups.loc[delay_groups[flag].eq("INVALID"), "total_rows"].sum()
    print(f"{flag} | null rows={null_rows:,} | invalid rows={invalid_rows:,}")

=== ARRIVAL DELAY MISSINGNESS ===
Source: C:\BigDataProject\output\format_benchmark\parquet
Rows: 5,819,079 | P0 match: True
Null ArrDelay: 105,071 | P0 match: True


,CANCELLED,DIVERTED,total_rows,null_arrdelay,nonnull_arrdelay,null_rate_pct,share_of_nulls_pct
0,0,0,5714008,0,5714008,0.0,0.00
1,0,1,15187,15187,0,100.0,14.45
2,1,0,89884,89884,0,100.0,85.55


CANCELLED | null rows=0 | invalid rows=0
DIVERTED | null rows=0 | invalid rows=0


#### **Arrival Delay Missingness: Results**

* The inspected Parquet output contains **5,819,079 rows** and **105,071 missing `ArrDelay` values**, corresponding to approximately **1.81%** of all records. Both totals match the corresponding counts recorded at P0.
* **89,884 records** have `CANCELLED = 1` and `DIVERTED = 0`. All have missing arrival delays, accounting for **85.55% of all missing `ArrDelay` values**.
* **15,187 records** have `CANCELLED = 0` and `DIVERTED = 1`. All have missing arrival delays, accounting for the remaining **14.45%**.
* The **5,714,008 records** with both flags equal to zero have **no missing `ArrDelay` values**. No records have both flags equal to one.
* Both status flags contain only **0/1 values**, with no null or out-of-domain values detected. In this inspected output, arrival-delay missingness aligns exactly with the recorded cancellation or diversion status.
* **The dataset and benchmark outputs remain unchanged.** Missing arrival delays are preserved, and this supplementary review addsevidence for their interpretation.

In [ ]:
from pathlib import Path

import pandas as pd
from IPython.display import display

lookup_path = Path(r"C:\BigDataProject\data\raw\airline\airlines.csv")
carrier_lookup = None

print("=== AIRLINE LOOKUP INSPECTION ===")

if not lookup_path.is_file():
    raise FileNotFoundError(f"Airline lookup not found: {lookup_path}")
else:
    carrier_lookup = pd.read_csv(
        lookup_path, dtype="string", keep_default_na=False
    )

    print(f"Source: {lookup_path}")
    print(f"Rows: {len(carrier_lookup):,}")
    print(f"Columns: {carrier_lookup.columns.tolist()}")

    display(carrier_lookup.head(20))

=== AIRLINE LOOKUP INSPECTION ===
Source: C:\BigDataProject\data\raw\airline\airlines.csv
Rows: 14
Columns: ['IATA_CODE', 'AIRLINE']


,IATA_CODE,AIRLINE
0,UA,United Air Lines Inc.
1,AA,American Airlines Inc.
2,US,US Airways Inc.
3,F9,Frontier Airlines Inc.
4,B6,JetBlue Airways
5,OO,Skywest Airlines Inc.
6,AS,Alaska Airlines Inc.
7,NK,Spirit Air Lines
8,WN,Southwest Airlines Co.
9,DL,Delta Air Lines Inc.


# **II. NYC Yellow Taxi 2024: Historical Dataset Audit**

## **1. Introduction & Objectives**

This section preserves the initial screening of **NYC TLC Yellow Taxi Trip Records 2024**, which was considered during the original candidate comparison. The inspected local snapshot consists of **12 monthly Parquet files** covering January–December 2024. **Airline 2015 remains the retained dataset for the completed benchmark and the agreed downstream experiments.**

- **Dataset:** Yellow Taxi Trip Records (Provider: NYC TLC)
- **Coverage:** January – December 2024 (12 files)
- **Native Format:** Parquet

**What this notebook executes:**
1. **Source Validation:** Verify that all 12 locally downloaded monthly files are available.
2. **Consolidation:** Verify schema consistency across all files and load them into a unified Spark DataFrame.
3. **Data Audit:** Inspect dataset dimensions (rows/columns), schema, missing values, and time range.
4. **Task 2 Compatibility Check:** Identify potential categorical/numeric fields for the query benchmark and assess the feasibility of extracting `Year` and `Month` from pickup timestamps.

**Scope Limitations:**
This notebook is strictly for **dataset validation and auditing**. It does **not** perform data cleaning, missing value imputation, column renaming, format conversion, or the actual benchmarking.

In [ ]:
# ============================================================
# NYC YELLOW TAXI 2024 DATASET AUDIT
# ============================================================

# ------------------------------------------------------------
# 1. Define dataset location and expected monthly files
# ------------------------------------------------------------

NYC_DATA_DIR = r"C:\BigDataProject\data\raw\nyc_taxi"

expected_files = [
    f"yellow_tripdata_2024-{month:02d}.parquet"
    for month in range(1, 13)
]

file_paths = [
    os.path.join(NYC_DATA_DIR, filename)
    for filename in expected_files
]

# ------------------------------------------------------------
# 2. Verify that all 12 monthly files exist
# ------------------------------------------------------------

print("=" * 70)
print("SOURCE FILE CHECK")
print("=" * 70)

missing_files = []

for path in file_paths:
    filename = os.path.basename(path)

    if os.path.exists(path):
        size_mb = os.path.getsize(path) / (1024 ** 2)
        print(f"[FOUND]   {filename:<35} {size_mb:>10.2f} MiB")
    else:
        print(f"[MISSING] {filename}")
        missing_files.append(filename)

if missing_files:
    raise FileNotFoundError(
        f"Missing files: {missing_files}"
    )

print("\nAll 12 monthly files are available.")

# ------------------------------------------------------------
# 3. Start Spark
# ------------------------------------------------------------

spark = (
    SparkSession.builder
    .appName("NYCYellowTaxi2024Audit")
    .master("local[*]")
    .config("spark.sql.shuffle.partitions", "8")
    .getOrCreate()
)

spark.sparkContext.setLogLevel("WARN")

# ------------------------------------------------------------
# 4. Check schema consistency across all 12 monthly files
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("SCHEMA CONSISTENCY CHECK")
print("=" * 70)

taxi_schemas = {}

for month, path in enumerate(file_paths, start=1):
    taxi_schemas[month] = spark.read.parquet(path).schema

taxi_reference_schema = taxi_schemas[1]

for month in range(1, 13):
    status = (
        "MATCH"
        if taxi_schemas[month] == taxi_reference_schema
        else "DIFFERENT"
    )

    print(f"2024-{month:02d}: {status}")

taxi_all_schema_match = all(
    schema == taxi_reference_schema
    for schema in taxi_schemas.values()
)

print(f"\nAll 12 taxi_schemas identical: {taxi_all_schema_match}")

# ------------------------------------------------------------
# 5. Read all 12 monthly files into one DataFrame
# ------------------------------------------------------------

taxi_df = spark.read.parquet(*file_paths)

print("\n" + "=" * 70)
print("BASIC DATASET INFORMATION")
print("=" * 70)

row_count = taxi_df.count()
column_count = len(taxi_df.columns)

print(f"Row count   : {row_count:,}")
print(f"Column count: {column_count}")

print("\nColumns:")
print(taxi_df.columns)

print("\nSchema:")
taxi_df.printSchema()

print("\nSample rows:")
taxi_df.show(5, truncate=False)

# ------------------------------------------------------------
# 6. Calculate missing values for all columns
# ------------------------------------------------------------

null_counts = (
    taxi_df.select([
        F.sum(F.col(c).isNull().cast("int")).alias(c)
        for c in taxi_df.columns
    ])
    .first()
    .asDict()
)

null_summary = spark.createDataFrame([
    (
        column,
        int(null_counts[column]),
        round(null_counts[column] / row_count * 100, 4)
    )
    for column in taxi_df.columns
], ["column", "null_count", "null_percentage"])

print("\n" + "=" * 70)
print("NULL VALUE SUMMARY")
print("=" * 70)

null_summary.orderBy(
    F.desc("null_percentage")
).show(
    column_count,
    truncate=False
)

# ------------------------------------------------------------
# 7. Check important columns for this candidate
# ------------------------------------------------------------

important_cols = [
    "VendorID",
    "tpep_pickup_datetime",
    "tpep_dropoff_datetime",
    "trip_distance",
    "total_amount"
]

print("\n" + "=" * 70)
print("IMPORTANT COLUMN CHECK")
print("=" * 70)

dtype_map = dict(taxi_df.dtypes)

for column in important_cols:
    if column in taxi_df.columns:
        print(f"[FOUND]   {column:<25} -> {dtype_map[column]}")
    else:
        print(f"[MISSING] {column}")

# ------------------------------------------------------------
# 8. Inspect important numeric and temporal fields
# ------------------------------------------------------------

summary = taxi_df.agg(
    F.min("tpep_pickup_datetime").alias("min_pickup"),
    F.max("tpep_pickup_datetime").alias("max_pickup"),

    F.countDistinct("VendorID").alias("distinct_vendors"),

    F.min("trip_distance").alias("min_trip_distance"),
    F.max("trip_distance").alias("max_trip_distance"),
    F.avg("trip_distance").alias("avg_trip_distance"),

    F.min("total_amount").alias("min_total_amount"),
    F.max("total_amount").alias("max_total_amount"),
    F.avg("total_amount").alias("avg_total_amount")
)

print("\n" + "=" * 70)
print("IMPORTANT FIELD STATISTICS")
print("=" * 70)

summary.show(truncate=False)

# ------------------------------------------------------------
# 9. Check whether YEAR and MONTH exist natively
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("YEAR / MONTH COMPATIBILITY")
print("=" * 70)

print(f"Native YEAR column : {'YES' if 'YEAR' in taxi_df.columns else 'NO'}")
print(f"Native MONTH column: {'YES' if 'MONTH' in taxi_df.columns else 'NO'}")

# ------------------------------------------------------------
# 10. Derive Year and Month temporarily for audit only
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("DERIVED YEAR / MONTH DISTRIBUTION")
print("=" * 70)

(
    taxi_df.groupBy(
        F.year("tpep_pickup_datetime").alias("YEAR"),
        F.month("tpep_pickup_datetime").alias("MONTH")
    )
    .count()
    .orderBy("YEAR", "MONTH")
    .show(50, truncate=False)
)

# ------------------------------------------------------------
# 11. Inspect VendorID distribution
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("VENDOR DISTRIBUTION")
print("=" * 70)

(
    taxi_df.groupBy("VendorID")
    .count()
    .orderBy(F.desc("count"))
    .show(20, truncate=False)
)

# ------------------------------------------------------------
# 12. Stop Spark
# ------------------------------------------------------------

spark.stop()

SOURCE FILE CHECK
[FOUND]   yellow_tripdata_2024-01.parquet          47.65 MB
[FOUND]   yellow_tripdata_2024-02.parquet          48.02 MB
[FOUND]   yellow_tripdata_2024-03.parquet          57.30 MB
[FOUND]   yellow_tripdata_2024-04.parquet          56.39 MB
[FOUND]   yellow_tripdata_2024-05.parquet          59.66 MB
[FOUND]   yellow_tripdata_2024-06.parquet          57.09 MB
[FOUND]   yellow_tripdata_2024-07.parquet          49.88 MB
[FOUND]   yellow_tripdata_2024-08.parquet          48.70 MB
[FOUND]   yellow_tripdata_2024-09.parquet          58.34 MB
[FOUND]   yellow_tripdata_2024-10.parquet          61.37 MB
[FOUND]   yellow_tripdata_2024-11.parquet          57.85 MB
[FOUND]   yellow_tripdata_2024-12.parquet          58.67 MB

All 12 monthly files are available.

SCHEMA CONSISTENCY CHECK
2024-01: MATCH
2024-02: MATCH
2024-03: MATCH
2024-04: MATCH
2024-05: MATCH
2024-06: MATCH
2024-07: MATCH
2024-08: MATCH
2024-09: MATCH
2024-10: MATCH
2024-11: MATCH
2024-12: MATCH

All 12 taxi_schema

## **Analysis of Results**

### **1. Source File Check**

* All 12 monthly files (January to December 2024) were successfully located. 
* The output clearly states `[FOUND]` for every month, confirming that Candidate B is correctly locked at the local snapshot level. 
* **Total files:** 12.
* **Size unit:** File sizes are expressed in **MiB**, calculated as `bytes / 1024**2`. The historical screening output uses the label MB for these values; its numerical values should be interpreted as MiB.

### **2. Schema Consistency**

- **Status:** `All 12 schemas identical: True`.
- Every monthly file from `2024-01` to `2024-12` returned a `MATCH` against the reference schema.
- **Conclusion:**The schemas of the 12 inspected monthly files matched the reference schema used in this screening.

### **3. Dataset Size**

- **Row count:** 41,169,720
- **Column count:** 19
- **Conclusion:** The inspected Taxi dataset contains 41,169,720 rows and 19 columns, compared with 5,819,079 rows and 31 columns in Airline. These dimensions describe the candidate workloads; the screening did not measure their relative benchmark performance.

### **4. Native Schema**

We verified the presence of key fields:
- `VendorID` (integer)
- `tpep_pickup_datetime` / `tpep_dropoff_datetime` (timestamp_ntz)
- `trip_distance` / `total_amount` (double)

The Taxi fields represent different concepts from Carrier and ArrDelay. The historical candidate comparison therefore identified a need for a taxi-specific grouping field and numeric measure.

### **5. Missing Values**

Based on the `NULL VALUE SUMMARY`:
- **Critical Columns:** `VendorID`, `tpep_pickup_datetime`, `trip_distance`, and `total_amount` have **0 missing values (100% complete)**.
- Secondary columns (e.g., `passenger_count`, `RatecodeID`, `congestion_surcharge`) have exactly 4,091,232 nulls (~9.94%).
- At this stage, we are only observing. No dropping or filling of nulls is performed.

### **6. Time Coverage**

- **min_pickup:** `2002-12-31 16:46:07`
- **max_pickup:** `2026-06-26 23:53:12`
- **Conclusion:** The inspected files contain pickup timestamps outside the nominal 2024 period. The screening records these values for interpretation; it does not establish their causes or apply a filtering rule.

### **7. Year and Month Availability**
- **Native YEAR column:** NO
- **Native MONTH column:** NO
- **Conclusion:** The Taxi dataset has no native Year or Month columns. A Year–Month partitioning experiment would require deriving these fields from tpep_pickup_datetime; the screening did not measure the cost of that preparation.

### **8. Derived Year/Month Distribution**

* **More than 99.9%** of the inspected records have pickup timestamps in **2024**.
* Small groups of records have pickup timestamps outside that period, including **2002, 2008, 2009, 2023, 2025, and 2026**.
* These groups are relevant when interpreting a potential Year–Month partitioned layout. **Their effect on output file counts and sizes was not measured in this screening.**

### **9. VendorID as a Categorical Variable**

- `VendorID` can technically be used for a categorical aggregate query (`GROUP BY VendorID`).
- However, the distribution is highly skewed: Vendor 2 dominates with ~31.4M rows, Vendor 1 has ~9.7M, and Vendors 6 & 7 have negligible counts. With only 4 distinct values, it is far less diverse than the 14 distinct carriers in the Airline dataset.

### **10. Numeric Measures**

Natural numeric metrics available are `trip_distance` and `total_amount`. Either can be used for aggregate queries like `AVG(trip_distance)`. 
Note: The recorded summaries include a minimum total_amount of −2,265 and a maximum trip_distance of 398,608. These values require interpretation in the dataset's context; the screening alone does not establish that every extreme value is invalid.

### **11. Candidate B Summary**

Following this audit, here is the final assessment of NYC Yellow Taxi 2024:

| Criteria | NYC Yellow Taxi 2024 |
|---|---|
| Full 12 monthly files present | Yes |
| 12 schemas identical | Yes |
| Row count | 41,169,720 |
| Column count | 19 |
| Has suitable categorical variable | Yes (`VendorID`), but heavily skewed (4 distinct) |
| Has suitable numeric variable | Yes (`trip_distance`, `total_amount`) |
| Has native Year | **No** |
| Has native Month | **No** |
| Can derive Year/Month | Yes (from `tpep_pickup_datetime`) |
| Natural equivalent to `Carrier` | No (requires semantic shift to `VendorID`) |
| Natural equivalent to `ArrDelay` | No (requires semantic shift to distance/amount) |
| Observations requiring interpretation | Pickup timestamps outside 2024 and recorded numeric extremes |
| Historical preparation considerations | Derive Year/Month from pickup timestamps and interpret the observed value ranges |

# **III. Dataset Comparison & Final Selection**

## **1. Objective**

The initial candidate comparison considered the following requirements of **Task 2: File Format Benchmarking and Partition Management**:
- Alignment with a query that groups by `Carrier` and computes `AVG(ArrDelay)`.
- Native support for the `.partitionBy("Year", "Month")` experiment.
- Minimal preprocessing overhead.
- Practicality for reproducible, multi-run local execution (warm-ups and iterations) to seamlessly integrate with ID4's workflow.

## **2. Candidate Summary**

| Criterion | Candidate A: Airline 2015 | Candidate B: NYC Yellow Taxi 2024 |
|---|---|---|
| **Source Structure** | 1 main CSV file | 12 monthly Parquet files |
| **Scale (Rows / Cols)** | 5,819,079 / 31 | 41,169,720 / 19 |
| **Native Categorical Field** | `AIRLINE` (14 distinct) | `VendorID` (4 distinct, heavily skewed) |
| **Native Numeric Measure** | `ARRIVAL_DELAY` | `trip_distance`, `total_amount` |
| **Native Year/Month Columns**| **Yes** | **No** (Derivation required) |
| **Semantic Match to Query** | **Exact match** (`Carrier`, `ArrDelay`) | **Requires query redesign** |
| **Partitioning preparation** | Native `YEAR` and `MONTH` columns | Derive Year/Month from pickup timestamps |
| **Data-quality treatment** | Original nulls retained under the benchmark contract | Out-of-year dates and numeric extremes documented during screening |

## **3. Comparison by Benchmark Requirements**

### **3.1 Read Query Compatibility**
The planned benchmark query groups records by `Carrier` and computes `AVG(ArrDelay)`.
- **Airline:** Maps natively without semantic changes (`AIRLINE` → `Carrier`, `ARRIVAL_DELAY` → `ArrDelay`).
- **Taxi:** Would require substituting the concepts entirely (e.g., `GROUP BY VendorID` and `AVG(trip_distance)`). 
**Query compatibility:** Airline provides the carrier and arrival-delay fields used by the retained benchmark.

### **3.2 Partition Control (`partitionBy`)**
* **Airline:** The recorded `YEAR` and `MONTH` fields contain no missing values and cover **2015** and all **12 months**. They provide the fields used for the planned Year/Month partitioning experiment.
* **Taxi:** Year and month must be derived from pickup timestamps. The screening identified sparse groups outside the intended 2024 period. These groups require attention when interpreting a partitioned layout; their impact on file counts and sizes was not measured in the screening.

### **3.3 Dataset Scale & Local Execution**
* **Airline:** The screened dataset contains **5,819,079 rows and 31 columns**. It provides the carrier and arrival-delay fields used by the benchmark, together with native year and month fields.
* **Taxi:** The screened dataset contains **41,169,720 rows and 19 columns**. It has a larger recorded row count, with different records, fields, and source-file characteristics.
**Selection rationale:** Airline was selected for its alignment with the required query and partitioning fields. The candidate-screening results do not constitute a controlled comparison of execution time between the two datasets.

### **3.4 Experimental Fairness (Note on Source Formats)**
* Airline and Taxi were screened from different source formats and contain different records and schemas. Their raw file sizes do not isolate the storage effect of CSV versus Parquet.
* The documented write protocol for the accepted Airline baseline uses the same canonical DataFrame across all four formats. It applies **`persist(StorageLevel.DISK_ONLY)` followed by `count()`** before the warm-up and measured writes.
* Materialization moves the initial source parsing and canonical preparation outside the measured write operations. **Accessing the persisted input remains part of each write job**, so the timings reflect a disk-persisted input configuration.
* Compression settings, fixed format order, and local runtime conditions remain relevant to the interpretation of the measured results.

## **4. Final Dataset Selection & Project Lock**

**The 2015 Flight Delays and Cancellations dataset is retained for the completed write benchmark and the agreed downstream experiments.**
The selection is supported by the existing carrier and arrival-delay fields, native year and month fields, and the recorded dataset profile. The benchmark results are interpreted under the documented protocol and the provenance limitations described in `REPORT.md`.

### **Canonical Schema Lock**
The accepted P0 dataset uses the following **four column renames**. This mapping defines the retained canonical field names for the ID3 write benchmark and the ID4 read and partitioning experiments.

| Original Column | Canonical Column | Data Type | Purpose |
|---|---|---|---|
| `AIRLINE` | `Carrier` | String | Categorical Grouping (`GROUP BY`) |
| `ARRIVAL_DELAY` | `ArrDelay` | Integer | Numeric Aggregation (`AVG`) |
| `YEAR` | `Year` | Integer | Directory Partitioning |
| `MONTH` | `Month` | Integer | Directory Partitioning |

**Retained data contract:** The benchmark preserves all **5,819,079 rows**, all **31 columns**, and the original null values. Canonicalization changes only the four column names listed above. No filtering, imputation, or deduplication is applied.

`DAY` and `DEPARTURE_DELAY` retain their original names and integer types. The supplementary checks document the quality and interpretation of the retained data.